### PR TITLE
JetStream consumer options

### PR DESCRIPTION
**Pull request by mtmk**
Implement JetStream Consumer API `consume()` options based on [ADR-37](https://github.com/nats-io/nats-architecture-and-design/blob/main/adr/ADR-37.md).

The design idea is to implement a `NatsJSSubBase` class to handle control messages and timeouts. Then we can implement specific classes for all three calls, `next()`, `fetch()` and `consume()`. In this PR only `consume()` is implemented.

### CODE REVIEW

**Comment by mtmk** [comment]
We seem to have an issue with `NATS.Client.Core.Tests.NatsConnectionTestTls.UserCredentialAuthTest` against nats-server dev version. Can't repro on my machine.

File: src/NATS.Client.JetStream/Internal/NatsNatsJSSubBaseConsume.cs [hunk 1]
```diff
@@ -0,0 +1,136 @@
...
+        // User messages are buffered here separately to allow smoother flow while control loop
+        // pulls more data in the background. This also allows control messages to be dealt with
+        // in the same loop as the control messages to keep state updates consistent. This is as
+        // opposed to having a control and a message channel at the point of serializing the messages
+        // in NatsJSSub class.
+        _userMessageChannel = Channel.CreateBounded<NatsJSMsg<T?>>(new BoundedChannelOptions(1_000)
+        {
+            FullMode = BoundedChannelFullMode.Wait,
+            AllowSynchronousContinuations = false,
+        });
```
Comment (caleblloyd) [inline src/NATS.Client.JetStream/Internal/NatsNatsJSSubBaseConsume.cs:82]: should this take the options from `NatsSubOpts.ChannelOptions`?

File: src/NATS.Client.JetStream/Internal/NatsNatsJSSubBaseConsume.cs [hunk 2]
```diff
@@ -0,0 +1,136 @@
...
+
+        // We drop the old message if notification handler isn't able to keep up.
+        // This is to avoid blocking the control loop and making sure we deliver all the messages.
+        // Assuming newer messages would be more relevant and worth keeping than older ones.
+        _notificationChannel = Channel.CreateBounded<NatsJSNotification>(new BoundedChannelOptions(1_000)
+        {
+            FullMode = BoundedChannelFullMode.DropOldest,
+            AllowSynchronousContinuations = false,
+        });
```
Comment (caleblloyd) [inline src/NATS.Client.JetStream/Internal/NatsNatsJSSubBaseConsume.cs:91]: 1k seems excessive but I don't really know what a good number would be - 10?  100? :laughing: 

Also doesn't really seem worth making configurable.  I guess they are small so 1k might be fine.

File: src/NATS.Client.JetStream/Internal/NatsNatsJSSubBaseConsume.cs (hunk 1 above)
Comment (mtmk) [reply src/NATS.Client.JetStream/Internal/NatsNatsJSSubBaseConsume.cs:82]: True 💯  I didn't think of that.

**Review by caleblloyd — APPROVED** [review]
LGTM

**Pull request merged**